### PR TITLE
Notes on random restarts, number of iterations and convergence

### PR DESCRIPTION
**Pull request by ftramer**
This PR addresses the points raised in https://github.com/evaluating-adversarial-robustness/adv-eval-paper/issues/18 and https://github.com/evaluating-adversarial-robustness/adv-eval-paper/issues/19.

Comments welcome!

### CODE REVIEW

**Comment by anishathalye** [comment]
I like the text you added on not imposing computational bounds on an attacker.

Do you think it's worth separating out iterations vs random restarts (line 765) into two separate bullet points? I think they are two ideas: one is about running to convergence, and the other is avoiding local optima.

**Comment by ftramer** [comment]
Thanks for the comment. I agree that these two points should be separated. The latest commit fixes this.

File: paper.tex [hunk 1]
```diff
@@ -1293,13 +1297,13 @@ \subsection{Verify Attack Convergence}
 have been found necessary \citep{qian2018l2}.
 
 There are few reasonable threat models under which an attacker can
-compute 100 iterations of gradient descent, but not 1000.
+compute 100 iterations of gradient descent, but not 1000. As stated in~\cite{athalye2018obfuscated}, merely increasing the time it takes an adversary to find an adversarial example (e.g., from one second to ten seconds) does not constitute an increase in robustness.
```
Comment (hendrycks) [inline paper.tex]: Increasing the time from 1 second to 10 days would constitute an increase in robustness, in the sense that the model is harder to break. The metric for some adversarial robustness competitions is the mean perturbation size required to fool a network, or the goal is to drive up the the distortion budget. Driving up the computational budget is also valuable.

File: paper.tex [hunk 2]
```diff
@@ -379,6 +379,8 @@ \subsubsection{Adversarial capabilities}
 hundred (or thousand) times from the distribution and computing the mean, such
 an approach is not possible for worst-case robustness.
 
+It is customary for defenses not to impose any computational bounds on an attacker (i.e., the above definitions of adversarial risk consider only the \emph{existence} of adversarial examples, and not the difficulty of finding them). We believe that restricting an adversary's computational power would only be interesting if we could base robustness on a cryptographic assumption (e.g., the model is robust under the assumption that the adversary cannot factor a product of large primes). We are not aware of any defense that achieves this. Stronger restrictions (e.g., the adversary is limited to attacks with $100$ iterations) are typically uninteresting in that they cannot be meaningfully enforced and there is usually no economic reason for an adversary not to spend a little more time to succeed.
```
Comment (hendrycks) [inline paper.tex]: > It is customary for defenses not to impose any computational bounds on an attacker (i.e., the above definitions of adversarial risk consider only the \emph{existence} of adversarial examples

Maksym Andriushchenko found a scenario in which an adversarial defense needed 1000 random restarts per example to bring down the model's accuracy to the fullest extent. Analysis by Synthesis also shows just how quickly the adversary's budget can swell---the forward pass can require 100000 samples per example. On an evaluation dataset of 10000 examples, and with a 100 step attacker, we are left with so many zeros that my head is left spinning. "Existence" is an appropriate aesthetic in maths where we can _choose_ any element we want without cost, but if there is not a good search procedure for that element, then it is possible for the problem to veer away from empirical study. But you probably think similarly given the comment about a cryptographic assumption.

File: paper.tex (hunk 1 above)
Comment (ftramer) [reply paper.tex]: This is a very tricky point to make, as it ends up depending on the computational model used. Computational cost of an attack is pretty much never taken into consideration (in any security context) unless either: (1) the cost can be argued to be exponential (e.g., for breaking crypto primitives); or (2) there is some clear economic model that explains when an attack is profitable or not.

File: paper.tex (hunk 2 above)
Comment (ftramer) [reply paper.tex]: Similar to my comment above, it is very tricky to say that robustness has really increased here. This is basically saying that the attacker just needs to find a faster attack and the model is easy to break. Unless there's a good reason to believe that optimizing the attack should be hard, this is a pretty weak degree of security, that is also very hard to evaluate or refute. E.g., if the difference between robustness and no robustness is whether the attacker took the time to write down the attack in assembly, run it on TPUs, and come up with a bunch of clever heuristics to squeeze out 1-2 orders of magnitude, that seems like an uninteresting direction to go for.

File: paper.tex (hunk 1 above)
Comment (hendrycks) [reply paper.tex]: > Computational cost of an attack is pretty much never taken into consideration (in any security context)

I agree. Computational cost is nearly always taken into consideration in machine learning contexts. I am not aware of a strong security motivation for \ell_p adversarial robustness, but I am aware of a machine learning motivation for adversarial robustness. Hence an author specifying that they limit the number of forward passes used by the adversary to 1 million per example would be fine by me.

Comment (ftramer) [reply paper.tex]: I don't quite understand your point here. 

The machine learning motivation for adversarial robustness is mainly about understanding why adversarial examples exist and how to remove them. What would be the machine learning motivation for training models that have no increased robustness but where finding adversarial examples is more computationally difficult? 

This only seems interesting from a security perspective (in which case I agree that something beyond \ell_p norms is needed). But in that case, you'd also need some security argument as to why an adversary wouldn't spend 10 days to find an adversarial example.

Comment (hendrycks) [reply paper.tex]: > The machine learning motivation for adversarial robustness is mainly about understanding why adversarial examples exist and how to remove them.

It is interesting because adversarial training can provide robustness of other sorts (such as robustness to stochastic corruptions irrespective of the model, other types of adversarial attacks such as adversarial weather (forthcoming), and naturally occurring classifier blindspots (forthcoming)). Adversarial training can also teach the network shape and not to rely on cheap textural cues. There is also an AI safety motivation based on Goodhart's law, but that's a separate discussion.

> you'd also need some security argument as to why an adversary wouldn't spend 10 days to find an adversarial example.

They of course would not spend 10 years, hence there is some security in jacking up the computational budget. While previous works should have increased the attacker budget by an order of magnitude, this is not a reason to suggest to dispense with computational concerns altogether.

Comment (ftramer) [reply paper.tex]: > They of course would not spend 10 years, hence there is some security in jacking up the computational budget.

I fully agree with this. But it is really hard to figure out what the "right" cutoff point should be. The only satisfactory way we've found so far (in crypto) is to say that increasing the attack cost by a polynomial factor is not enough.

I'm worried that if the community agrees that increasing the attacker's cost is a valuable contribution, we'll see a whole bunch of defenses make arguments like "we've increased the cost of attack by 100x or 1000x". First, this is a very difficult argument to make correctly (i.e., just because your implementation of the attack is 100x slower on your platform doesn't mean this is inherent), and second, it is hard to say whether this is really meaningful or not without context.

> It is interesting because adversarial training can provide robustness of other sorts (such as robustness to stochastic corruptions irrespective of the model, other types of adversarial attacks such as adversarial weather (forthcoming), and naturally occurring classifier blindspots (forthcoming)). Adversarial training can also teach the network shape and not to rely on cheap textural cues. There is also an AI safety motivation based on Goodhart's law, but that's a separate discussion.

These points I all agree with. But note that for all of these, the cost of the "attack" doesn't really matter (e.g., if you use adversarial training to get better robustness to rare weather conditions, you still care about the existence of adversarial conditions rather than the computational cost of finding them).

File: paper.tex (hunk 2 above)
Comment (hendrycks) [reply paper.tex]: > has _really_ increased here (emphasis mine)

I am not claiming adversarial robustness has categorically increased in that situation, but I am claiming that computational concerns can be relevant.

> Unless there's a good reason to believe that optimizing the attack should be hard

This could happen, in which case computation is relevant. Of course now it's not currently relevant since attacks are cheap. But if this document is forward-looking, then saying a factor as often important as computation has no relevance seems to exclusively prioritize security precedents and dismiss machine learning precedents. If the document is not trying to be forward-looking, then I absolutely agree computational cost is usually not relevant in evaluation and probably won't be for minor variations on current defenses and attacks. 

Rather than making tough calls about the irrelevance of computation, perhaps just recommend increasing the attacker budget in various respects by an order of magnitude to the point of convergence, as the rest of the document currently does? Thank you for the fast replies.

File: paper.tex (hunk 1 above)
Comment (hendrycks) [reply paper.tex]: > we've increased the cost of attack by 100x or 1000x

Defense papers ought not choose to make the attacker weaker than the baseline attacker.

> you still care about the existence of adversarial conditions rather than the computational cost of finding them

If the computational cost is in any way tied to the frequency of occurrence of such conditions (imagine that an adversarial distortion is highly influenced by the initialization point), then computational cost becomes relevant. (I should note that all my responses so far are colored by a hypothetical, speculated defense in which any attacker with a fixed distortion budget is as successful as an attacker randomly sampling perturbations within the same distortion budget.)

File: paper.tex (hunk 2 above)
Comment (ftramer) [reply paper.tex]: I don't think we want to say that computation cost is never relevant.
With the paragraph I added in line 382, I wanted to say that this starts to be relevant once you can formally say that attacking is computationally expensive, or if you can argue that the economic incentives at play make attacking too costly.
To me, this is forward-looking in that we say that right now, computational cost isn't interesting to look at because we really can't say anything meaningful, but if someone writes a paper some day that does make significant progress along these lines, that would be great.

> Rather than making tough calls about the irrelevance of computation, perhaps just recommend increasing the attacker budget in various respects by an order of magnitude to the point of convergence, as the rest of the document currently does? 

The document already says this for the number of iterations, which is a proxy for the computation time. Did you have something else in mind?

File: paper.tex (hunk 1 above)
Comment (ftramer) [reply paper.tex]: > If the computational cost is in any way tied to the frequency of occurrence of such conditions (imagine that an adversarial distortion is highly influenced by the initialization point), then computational cost becomes relevant.

I agree with this, and if you can argue that this is true in a specific setting, that might be fine. The big difference here is that you would be tying the computational cost to a specific measurable quantity (e.g., the success-rate of a random attacker as in your speculative example).

File: paper.tex (hunk 2 above)
Comment (hendrycks) [reply paper.tex]: > So the document already say this

I acknowledge the document says this for iterations. It should probably say something similar for the number of random restarts as well.

The text that I took issue with was as follows.

> merely increasing the time it takes an adversary to find an adversarial example... does not constitute an increase in robustness.

Perhaps instead

> merely causing an adversary to require an order of magnitude more time to create an adversarial example often does not constitute an increase in adversarial robustness.

Comment (ftramer) [reply paper.tex]: > merely increasing the time it takes an adversary to find an adversarial example... does not constitute an increase in robustness.

Ah I see, yes stated like this it is a bit too strict. I'll rephrase this to something more along the lines of line 382.

Comment (hendrycks) [reply paper.tex]: I'm glad to see we're not really in disagreement. Thanks!

**Comment by hendrycks** [comment]
Nit: https://arxiv.org/pdf/1810.12042.pdf should probably be cited since it speaks much about the importance of restarts in evaluation. This is demonstrated well in Table 3.

**Comment by ftramer** [comment]
That paper is cited in §4.8 and in §4.9. Was there another place where you think a citation is appropriate?

**Comment by hendrycks** [comment]
Oops, I'm a bungler. I was looking for Maksym's last name in the citations you added, but it's mosbach2018logit. Everything looks good.

**Comment by anishathalye** [comment]
Does anyone else have any feedback? If not, shall we go ahead and merge this PR?

**Comment by jonasrauber** [comment]
The proposed changes look good to me.

**Comment by wielandbrendel** [comment]
I definitely agree with the proposed change, and the point is formulated nicely.

**Comment by anishathalye** [comment]
Great! Merged.
